File: artifacts/pos-desktop/src/lib/customers.ts
// Local customers catalog — mirrors items.ts pattern.
// Browser mode: localStorage. Tauri mode: SQLite via Rust commands
// (customers::list_customers / upsert_customers / create_customer).

import { LS_KEYS, lsRead, lsWrite } from "./localStore";
// Task #207: shared-data Rust commands route through the bridge (a LAN
// client forwards them to the host; single/host call the local command).
import { bridgeInvoke as tauriInvoke, shouldUseBridge } from "./bridge";
import { enqueuePush } from "./pushQueue";

export interface LocalCustomer {
  id: number;
  cloudId?: number | null;
  nameAr: string;
  nameEn?: string | null;
  phone?: string | null;
  vatNumber?: string | null;
  currencyCode?: string;
  /** Shadow balance in base currency. Positive = customer owes us (مدين). */
  balance?: number;
  /** Credit control — max outstanding AR allowed (0 = unlimited). */
  creditLimit?: number;
  /** When true, a credit sale that exceeds creditLimit is rejected. */
  enforceCreditLimit?: boolean;
  /** Grace days before an unpaid credit invoice is overdue (0 = no check). */
  paymentTermsDays?: number;
  // ── Profile parity with web (Phase 1A) ──
  crNumber?: string | null;
  email?: string | null;
  city?: string | null;
  district?: string | null;
  street?: string | null;
  buildingNumber?: string | null;
  postalCode?: string | null;
  country?: string | null;
  nationalAddressShort?: string | null;
  locationLat?: string | null;
  locationLng?: string | null;
  locationLink?: string | null;
  /** When false, excluded from per-customer AR statement/ageing reports. */
  includeInStatements?: boolean;
  /** Default home branch (FK to branches_local.id). */
  branchId?: number | null;
  /** Dedicated GL receivable account (FK to accounts_local.id). null = default POS AR. */
  accountId?: number | null;
  createdAt?: string;
  updatedAt?: string;
  /** Soft-delete tombstone (overlay). When true, listCustomers filters this row out. */
  deleted?: boolean;
}

interface RustCustomer {
  id: number;
  cloud_id: number | null;
  name_ar: string;
  name_en: string | null;
  phone: string | null;
  vat_number: string | null;
  updated_at: string | null;
  currency_code?: string;
  balance?: number;
  credit_limit?: number;
  enforce_credit_limit?: boolean;
  payment_terms_days?: number;
  cr_number?: string | null;
  email?: string | null;
  city?: string | null;
  district?: string | null;
  street?: string | null;
  building_number?: string | null;
  postal_code?: string | null;
  country?: string | null;
  national_address_short?: string | null;
  location_lat?: string | null;
  location_lng?: string | null;
  location_link?: string | null;
  include_in_statements?: boolean;
  branch_id?: number | null;
  account_id?: number | null;
}

function fromRust(r: RustCustomer): LocalCustomer {
  return {
    id: r.id,
    cloudId: r.cloud_id,
    nameAr: r.name_ar,
    nameEn: r.name_en,
    phone: r.phone,
    vatNumber: r.vat_number,
    currencyCode: r.currency_code ?? "SAR",
    balance: r.balance ?? 0,
    creditLimit: r.credit_limit ?? 0,
    enforceCreditLimit: r.enforce_credit_limit ?? false,
    paymentTermsDays: r.payment_terms_days ?? 0,
    crNumber: r.cr_number ?? null,
    email: r.email ?? null,
    city: r.city ?? null,
    district: r.district ?? null,
    street: r.street ?? null,
    buildingNumber: r.building_number ?? null,
    postalCode: r.postal_code ?? null,
    country: r.country ?? null,
    nationalAddressShort: r.national_address_short ?? null,
    locationLat: r.location_lat ?? null,
    locationLng: r.location_lng ?? null,
    locationLink: r.location_link ?? null,
    includeInStatements: r.include_in_statements ?? true,
    branchId: r.branch_id ?? null,
    accountId: r.account_id ?? null,
    updatedAt: r.updated_at ?? undefined,
  };
}

export async function listCustomers(search?: string): Promise<LocalCustomer[]> {
  // MERGE strategy (mirrors items.ts): read both SQLite and localStorage.
  // updateCustomer/deleteCustomer write to localStorage only, so the merged
  // view is what makes those edits visible after the change.
  const fromTauri: LocalCustomer[] = [];
  if (shouldUseBridge()) {
    try {
      const rows = await tauriInvoke<RustCustomer[]>("list_customers", { search: search ?? null });
      fromTauri.push(...rows.map(fromRust));
    } catch { /* fall through */ }
  }
  const fromLs = lsRead<LocalCustomer[]>(LS_KEYS.customers, []);
  // OVERLAY: LS rows that match a Tauri row (by id OR cloudId) SUPERSEDE
  // the Tauri version — this is how updateCustomer/deleteCustomer can
  // mutate cloud-pulled rows without Rust write commands.
  const lsById = new Map<number, LocalCustomer>();
  const lsByCloud = new Map<number, LocalCustomer>();
  for (const r of fromLs) {
    lsById.set(r.id, r);
    if (r.cloudId) lsByCloud.set(r.cloudId, r);
  }
  const usedLs = new Set<number>();
  const merged: LocalCustomer[] = [];
  for (const t of fromTauri) {
    const overlay = lsById.get(t.id) ?? (t.cloudId ? lsByCloud.get(t.cloudId) : undefined);
    if (overlay) {
      merged.push(overlay);
      usedLs.add(overlay.id);
    } else {
      merged.push(t);
    }
  }
  for (const r of fromLs) {
    if (usedLs.has(r.id)) continue;
    merged.push(r);
  }
  // Filter tombstones (deleted rows the user removed locally).
  const visible = merged.filter((c) => !c.deleted);
  if (!search) return visible;
  const q2 = search.toLowerCase();
  return visible.filter((c) =>
    c.nameAr.includes(search) ||
    (c.nameEn ?? "").toLowerCase().includes(q2) ||
    (c.phone ?? "").includes(search) ||
    (c.vatNumber ?? "").includes(search),
  );
}

export async function getCustomer(id: number): Promise<LocalCustomer | null> {
  const all = await listCustomers();
  return all.find((c) => c.id === id) ?? null;
}

// Bulk upsert from cloud Pull. Matches on cloudId.
export async function upsertCustomersFromCloud(remote: Array<{
  id: number; nameAr: string; nameEn: string | null; phone: string | null;
  vatNumber: string | null; createdAt?: string;
}>): Promise<number> {
  if (shouldUseBridge()) {
    try {
      return await tauriInvoke<number>("upsert_customers_from_cloud", {
        rows: remote.map((c) => ({
          cloud_id: c.id,
          name_ar: c.nameAr,
          name_en: c.nameEn,
          phone: c.phone,
          vat_number: c.vatNumber,
        })),
      });
    } catch {
      // fall through
    }
  }
  const existing = lsRead<LocalCustomer[]>(LS_KEYS.customers, []);
  const byCloud = new Map<number, LocalCustomer>();
  for (const c of existing) if (c.cloudId) byCloud.set(c.cloudId, c);
  let maxId = existing.reduce((m, c) => Math.max(m, c.id), 0);

  for (const r of remote) {
    const prev = byCloud.get(r.id);
    if (prev) {
      Object.assign(prev, {
        nameAr: r.nameAr,
        nameEn: r.nameEn,
        phone: r.phone,
        vatNumber: r.vatNumber,
        updatedAt: new Date().toISOString(),
      });
    } else {
      maxId += 1;
      existing.push({
        id: maxId,
        cloudId: r.id,
        nameAr: r.nameAr,
        nameEn: r.nameEn,
        phone: r.phone,
        vatNumber: r.vatNumber,
        createdAt: r.createdAt,
        updatedAt: new Date().toISOString(),
      });
    }
  }
  lsWrite(LS_KEYS.customers, existing);
  return remote.length;
}

export interface CreateCustomerInput {
  nameAr: string;
  nameEn?: string | null;
  phone?: string | null;
  vatNumber?: string | null;
  currencyCode?: string;
  /** Opening balance amount (native, > 0). Posted as a JE on create only. */
  openingBalance?: number;
  /** "debit" (مدين — owes us) or "credit" (دائن — we owe them). */
  openingNature?: "debit" | "credit";
  openingDate?: string;
  /** Credit control. */
  creditLimit?: number;
  enforceCreditLimit?: boolean;
  paymentTermsDays?: number;
  // ── Profile parity with web (Phase 1A) ──
  crNumber?: string | null;
  email?: string | null;
  city?: string | null;
  district?: string | null;
  street?: string | null;
  buildingNumber?: string | null;
  postalCode?: string | null;
  country?: string | null;
  nationalAddressShort?: string | null;
  locationLat?: string | null;
  locationLng?: string | null;
  locationLink?: string | null;
  includeInStatements?: boolean;
  branchId?: number | null;
  accountId?: number | null;
}

/** Build the Rust `profile` struct arg from a customer input (camelCase keys). */
function toProfile(input: CreateCustomerInput) {
  return {
    crNumber: input.crNumber ?? null,
    email: input.email ?? null,
    city: input.city ?? null,
    district: input.district ?? null,
    street: input.street ?? null,
    buildingNumber: input.buildingNumber ?? null,
    postalCode: input.postalCode ?? null,
    country: input.country ?? null,
    nationalAddressShort: input.nationalAddressShort ?? null,
    locationLat: input.locationLat ?? null,
    locationLng: input.locationLng ?? null,
    locationLink: input.locationLink ?? null,
    includeInStatements: input.includeInStatements ?? null,
    // branch_id tri-state on the Rust wire: undefined → null (preserve via
    // COALESCE), explicit null ("— بدون —") → 0 sentinel (clear to NULL),
    // a real id → that id. SQLite branch ids start at 1, so 0 is never valid.
    branchId: input.branchId === null ? 0 : (input.branchId ?? null),
    // Same tri-state as branchId: undefined → null (preserve via COALESCE/CASE),
    // explicit null ("— بدون —") → 0 sentinel (clear to NULL), real id → that id.
    accountId: input.accountId === null ? 0 : (input.accountId ?? null),
  };
}

export async function createCustomer(input: CreateCustomerInput): Promise<LocalCustomer> {
  const now = new Date().toISOString();
  let created: LocalCustomer;
  if (shouldUseBridge()) {
    try {
      const r = await tauriInvoke<RustCustomer>("create_customer_local", {
        nameAr: input.nameAr,
        nameEn: input.nameEn ?? null,
        phone: input.phone ?? null,
        vatNumber: input.vatNumber ?? null,
        currencyCode: input.currencyCode ?? "SAR",
        openingBalance: input.openingBalance ?? null,
        openingNature: input.openingNature ?? null,
        openingDate: input.openingDate ?? null,
        creditLimit: input.creditLimit ?? null,
        enforceCreditLimit: input.enforceCreditLimit ?? null,
        paymentTermsDays: input.paymentTermsDays ?? null,
        profile: toProfile(input),
      });
      created = fromRust(r);
    } catch {
      created = createInLocalStorage(input, now);
    }
  } else {
    created = createInLocalStorage(input, now);
  }
  // Queue for cloud push (idempotent via clientId).
  enqueuePush({
    clientId: `cust-${created.id}-${Date.now()}`,
    entityType: "customer",
    operation: "create",
    payload: {
      localId: created.id,
      nameAr: created.nameAr,
      nameEn: created.nameEn,
      phone: created.phone,
      vatNumber: created.vatNumber,
    },
  });
  return created;
}

function createInLocalStorage(input: CreateCustomerInput, now: string): LocalCustomer {
  const all = lsRead<LocalCustomer[]>(LS_KEYS.customers, []);
  const id = all.reduce((m, c) => Math.max(m, c.id), 0) + 1;
  const row: LocalCustomer = {
    id,
    cloudId: null,
    nameAr: input.nameAr,
    nameEn: input.nameEn ?? null,
    phone: input.phone ?? null,
    vatNumber: input.vatNumber ?? null,
    currencyCode: input.currencyCode ?? "SAR",
    balance: 0,
    creditLimit: input.creditLimit ?? 0,
    enforceCreditLimit: input.enforceCreditLimit ?? false,
    paymentTermsDays: input.paymentTermsDays ?? 0,
    crNumber: input.crNumber ?? null,
    email: input.email ?? null,
    city: input.city ?? null,
    district: input.district ?? null,
    street: input.street ?? null,
    buildingNumber: input.buildingNumber ?? null,
    postalCode: input.postalCode ?? null,
    country: input.country ?? null,
    nationalAddressShort: input.nationalAddressShort ?? null,
    locationLat: input.locationLat ?? null,
    locationLng: input.locationLng ?? null,
    locationLink: input.locationLink ?? null,
    includeInStatements: input.includeInStatements ?? true,
    branchId: input.branchId ?? null,
    accountId: input.accountId ?? null,
    createdAt: now,
    updatedAt: now,
  };
  all.push(row);
  lsWrite(LS_KEYS.customers, all);
  return row;
}

export async function updateCustomer(id: number, patch: CreateCustomerInput): Promise<LocalCustomer | null> {
  // Persist to SQLite first so the Rust-side credit-limit enforcement on sales
  // reads the true values (the LS overlay below is invisible to Rust). Browser
  // mode and any bridge failure fall through to the LS-overlay path.
  if (shouldUseBridge()) {
    try {
      await tauriInvoke<RustCustomer>("update_customer_local", {
        id,
        nameAr: patch.nameAr ?? null,
        nameEn: patch.nameEn ?? null,
        phone: patch.phone ?? null,
        vatNumber: patch.vatNumber ?? null,
        currencyCode: patch.currencyCode ?? null,
        creditLimit: patch.creditLimit ?? null,
        enforceCreditLimit: patch.enforceCreditLimit ?? null,
        paymentTermsDays: patch.paymentTermsDays ?? null,
        profile: toProfile(patch),
      });
    } catch { /* fall through to LS overlay */ }
  }
  const all = lsRead<LocalCustomer[]>(LS_KEYS.customers, []);
  const idx = all.findIndex((c) => c.id === id);
  let updated: LocalCustomer;
  if (idx >= 0) {
    updated = { ...all[idx], ...patch, updatedAt: new Date().toISOString() };
    all[idx] = updated;
  } else {
    // Row originates from SQLite — write a full overlay to LS so the
    // edit becomes visible and the next update can find it.
    const merged = await listCustomers();
    const base = merged.find((c) => c.id === id);
    if (!base) return null;
    updated = { ...base, ...patch, updatedAt: new Date().toISOString() };
    all.push(updated);
  }
  lsWrite(LS_KEYS.customers, all);
  enqueuePush({
    clientId: `cust-upd-${id}-${Date.now()}`,
    entityType: "customer",
    operation: "update",
    payload: { localId: id, cloudId: updated.cloudId, ...patch },
  });
  return updated;
}

export async function deleteCustomer(id: number): Promise<void> {
  // Tombstone strategy — same as deleteItem. Cloud-backed rows get a
  // `deleted:true` overlay so the merged listCustomers hides them even
  // when SQLite still has the original; pure-local rows are dropped.
  const all = lsRead<LocalCustomer[]>(LS_KEYS.customers, []);
  const idx = all.findIndex((c) => c.id === id);
  let cloudId: number | null = null;
  if (idx >= 0) {
    cloudId = all[idx].cloudId ?? null;
    if (cloudId) {
      all[idx] = { ...all[idx], deleted: true, updatedAt: new Date().toISOString() };
    } else {
      all.splice(idx, 1);
    }
  } else {
    const merged = await listCustomers();
    const base = merged.find((c) => c.id === id);
    if (!base) return;
    cloudId = base.cloudId ?? null;
    all.push({ ...base, deleted: true, updatedAt: new Date().toISOString() });
  }
  lsWrite(LS_KEYS.customers, all);
  enqueuePush({
    clientId: `cust-del-${id}-${Date.now()}`,
    entityType: "customer",
    operation: "delete",
    payload: { localId: id, cloudId },
  });
}

export async function countCustomers(): Promise<number> {
  return (await listCustomers()).length;
}

File: artifacts/pos-desktop/src-tauri/src/customers.rs
// Local customers catalog backed by SQLite (`customers_local` table).
//
// Three read/write paths consumed by JS-side lib/customers.ts:
//   - list_customers          : full catalog for the admin grid
//   - upsert_customers_from_cloud : batched insert/update from /api/sync/pull
//   - create_customer_local   : cashier-created row (cloud_id NULL until push)
//
// The cloud sync key is `cloud_id` (UNIQUE in db.rs). A locally-created
// customer has cloud_id=NULL; the JS-side push queue eventually assigns
// one after a successful /api/sync/push response.

use crate::db;
use anyhow::Result;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct LocalCustomer {
    pub id: i64,
    pub cloud_id: Option<i64>,
    pub name_ar: String,
    pub name_en: Option<String>,
    pub phone: Option<String>,
    pub vat_number: Option<String>,
    pub updated_at: Option<String>,
    pub currency_code: String,
    pub balance: f64,
    pub credit_limit: f64,
    pub enforce_credit_limit: bool,
    pub payment_terms_days: i64,
    // ── Profile parity with web (Phase 1A) ──
    pub cr_number: Option<String>,
    pub email: Option<String>,
    pub city: Option<String>,
    pub district: Option<String>,
    pub street: Option<String>,
    pub building_number: Option<String>,
    pub postal_code: Option<String>,
    pub country: Option<String>,
    pub national_address_short: Option<String>,
    pub location_lat: Option<String>,
    pub location_lng: Option<String>,
    pub location_link: Option<String>,
    pub include_in_statements: bool,
    pub branch_id: Option<i64>,
}

const MAX_ROWS: i64 = 500;

// Shared SELECT column list — keep order in lockstep with `row_to_customer`.
const SELECT_COLS: &str = "id, cloud_id, name_ar, name_en, phone, vat_number, updated_at, currency_code, balance, \
     credit_limit, enforce_credit_limit, payment_terms_days, \
     cr_number, email, city, district, street, building_number, postal_code, country, \
     national_address_short, location_lat, location_lng, location_link, include_in_statements, branch_id";

fn row_to_customer(row: &rusqlite::Row<'_>) -> rusqlite::Result<LocalCustomer> {
    Ok(LocalCustomer {
        id: row.get(0)?,
        cloud_id: row.get(1)?,
        name_ar: row.get(2)?,
        name_en: row.get(3)?,
        phone: row.get(4)?,
        vat_number: row.get(5)?,
        updated_at: row.get(6)?,
        currency_code: row.get::<_, Option<String>>(7)?.unwrap_or_else(|| "SAR".to_string()),
        balance: row.get(8)?,
        credit_limit: row.get::<_, Option<f64>>(9)?.unwrap_or(0.0),
        enforce_credit_limit: row.get::<_, Option<i64>>(10)?.unwrap_or(0) != 0,
        payment_terms_days: row.get::<_, Option<i64>>(11)?.unwrap_or(0),
        cr_number: row.get(12)?,
        email: row.get(13)?,
        city: row.get(14)?,
        district: row.get(15)?,
        street: row.get(16)?,
        building_number: row.get(17)?,
        postal_code: row.get(18)?,
        country: row.get(19)?,
        national_address_short: row.get(20)?,
        location_lat: row.get(21)?,
        location_lng: row.get(22)?,
        location_link: row.get(23)?,
        include_in_statements: row.get::<_, Option<i64>>(24)?.unwrap_or(1) != 0,
        branch_id: row.get(25)?,
    })
}

#[tauri::command]
pub fn list_customers(search: Option<String>) -> Result<Vec<LocalCustomer>, String> {
    list(search.as_deref()).map_err(|e| e.to_string())
}

fn list(search: Option<&str>) -> Result<Vec<LocalCustomer>> {
    let conn = db::open()?;
    let sql = format!(
        "SELECT {SELECT_COLS}
               FROM customers_local
               WHERE (?1 IS NULL OR name_ar LIKE ?2 OR phone LIKE ?2 OR vat_number LIKE ?2)
               ORDER BY name_ar
               LIMIT ?3"
    );
    let pattern = search.map(|s| format!("%{}%", s));
    let mut stmt = conn.prepare(&sql)?;
    let rows = stmt.query_map(
        rusqlite::params![search, pattern, MAX_ROWS],
        row_to_customer,
    )?;
    let mut out = Vec::new();
    for r in rows { out.push(r?); }
    Ok(out)
}

#[derive(Deserialize, Debug)]
pub struct CloudCustomer {
    pub cloud_id: i64,
    pub name_ar: String,
    pub name_en: Option<String>,
    pub phone: Option<String>,
    pub vat_number: Option<String>,
}

#[tauri::command]
pub fn upsert_customers_from_cloud(rows: Vec<CloudCustomer>) -> Result<u64, String> {
    upsert_from_cloud(rows).map_err(|e| e.to_string())
}

fn upsert_from_cloud(rows: Vec<CloudCustomer>) -> Result<u64> {
    let mut conn = db::open()?;
    let tx = conn.transaction_with_behavior(rusqlite::TransactionBehavior::Immediate)?;
    let mut stmt = tx.prepare(
        "INSERT INTO customers_local
           (cloud_id, name_ar, name_en, phone, vat_number, updated_at)
         VALUES (?1, ?2, ?3, ?4, ?5, CURRENT_TIMESTAMP)
         ON CONFLICT(cloud_id) DO UPDATE SET
           name_ar = excluded.name_ar,
           name_en = excluded.name_en,
           phone = excluded.phone,
           vat_number = excluded.vat_number,
           updated_at = CURRENT_TIMESTAMP",
    )?;
    let mut count = 0_u64;
    for r in &rows {
        stmt.execute(rusqlite::params![
            r.cloud_id, r.name_ar, r.name_en, r.phone, r.vat_number,
        ])?;
        count += 1;
    }
    drop(stmt);
    tx.commit()?;
    Ok(count)
}

/// Optional profile fields shared by create + update. Bundling them in one
/// struct keeps the Tauri command signatures sane as the customer model grows.
/// JS passes a single `profile` object (camelCase keys → serde rename).
#[derive(Deserialize, Debug, Default, Clone)]
#[serde(rename_all = "camelCase")]
pub struct CustomerProfile {
    pub cr_number: Option<String>,
    pub email: Option<String>,
    pub city: Option<String>,
    pub district: Option<String>,
    pub street: Option<String>,
    pub building_number: Option<String>,
    pub postal_code: Option<String>,
    pub country: Option<String>,
    pub national_address_short: Option<String>,
    pub location_lat: Option<String>,
    pub location_lng: Option<String>,
    pub location_link: Option<String>,
    pub include_in_statements: Option<bool>,
    pub branch_id: Option<i64>,
}

#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub fn create_customer_local(
    name_ar: String,
    name_en: Option<String>,
    phone: Option<String>,
    vat_number: Option<String>,
    currency_code: Option<String>,
    opening_balance: Option<f64>,
    opening_nature: Option<String>,
    opening_date: Option<String>,
    credit_limit: Option<f64>,
    enforce_credit_limit: Option<bool>,
    payment_terms_days: Option<i64>,
    profile: Option<CustomerProfile>,
) -> Result<LocalCustomer, String> {
    create(name_ar, name_en, phone, vat_number, currency_code, opening_balance, opening_nature, opening_date,
           credit_limit, enforce_credit_limit, payment_terms_days, profile.unwrap_or_default())
        .map_err(|e| e.to_string())
}

#[allow(clippy::too_many_arguments)]
fn create(
    name_ar: String,
    name_en: Option<String>,
    phone: Option<String>,
    vat_number: Option<String>,
    currency_code: Option<String>,
    opening_balance: Option<f64>,
    opening_nature: Option<String>,
    opening_date: Option<String>,
    credit_limit: Option<f64>,
    enforce_credit_limit: Option<bool>,
    payment_terms_days: Option<i64>,
    profile: CustomerProfile,
) -> Result<LocalCustomer> {
    let mut conn = db::open()?;
    let cur = currency_code.unwrap_or_else(|| "SAR".to_string());
    let cl = credit_limit.unwrap_or(0.0);
    let enforce = enforce_credit_limit.unwrap_or(false);
    let terms = payment_terms_days.unwrap_or(0);
    let country = profile.country.clone().unwrap_or_else(|| "SA".to_string());
    let include = profile.include_in_statements.unwrap_or(true);
    // 0 is the "no branch" sentinel from the JS wire (see toProfile); SQLite
    // branch ids start at 1, so collapse it to NULL on insert.
    let branch_id = match profile.branch_id {
        Some(0) => None,
        other => other,
    };
    let tx = conn.transaction()?;
    tx.execute(
        "INSERT INTO customers_local (cloud_id, name_ar, name_en, phone, vat_number, currency_code,
                                      credit_limit, enforce_credit_limit, payment_terms_days,
                                      cr_number, email, city, district, street, building_number, postal_code,
                                      country, national_address_short, location_lat, location_lng, location_link,
                                      include_in_statements, branch_id, updated_at)
         VALUES (NULL, ?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8,
                 ?9, ?10, ?11, ?12, ?13, ?14, ?15,
                 ?16, ?17, ?18, ?19, ?20,
                 ?21, ?22, CURRENT_TIMESTAMP)",
        rusqlite::params![
            name_ar, name_en, phone, vat_number, cur, cl, enforce as i64, terms,
            profile.cr_number, profile.email, profile.city, profile.district, profile.street,
            profile.building_number, profile.postal_code, country, profile.national_address_short,
            profile.location_lat, profile.location_lng, profile.location_link,
            include as i64, branch_id,
        ],
    )?;
    let id = tx.last_insert_rowid();
    let ob = opening_balance.unwrap_or(0.0).abs();
    if ob > 1e-9 {
        let nature = opening_nature.as_deref().unwrap_or("debit");
        let date = match opening_date {
            Some(d) if !d.trim().is_empty() => d,
            _ => tx.query_row("SELECT date('now','localtime')", [], |r| r.get(0))?,
        };
        crate::accounting::post_party_opening_balance(&tx, "customer", id, &cur, ob, nature, &date)?;
    }
    let balance: f64 = tx.query_row("SELECT balance FROM customers_local WHERE id=?1", rusqlite::params![id], |r| r.get(0))?;
    tx.commit()?;
    Ok(LocalCustomer {
        id,
        cloud_id: None,
        name_ar,
        name_en,
        phone,
        vat_number,
        updated_at: None,
        currency_code: cur,
        balance,
        credit_limit: cl,
        enforce_credit_limit: enforce,
        payment_terms_days: terms,
        cr_number: profile.cr_number,
        email: profile.email,
        city: profile.city,
        district: profile.district,
        street: profile.street,
        building_number: profile.building_number,
        postal_code: profile.postal_code,
        country: Some(country),
        national_address_short: profile.national_address_short,
        location_lat: profile.location_lat,
        location_lng: profile.location_lng,
        location_link: profile.location_link,
        include_in_statements: include,
        branch_id,
    })
}

/// Persist editable scalar fields (incl. credit control) to SQLite. Unlike the
/// LS-overlay path used by the browser fallback, this writes the canonical row
/// so the Rust-side credit-limit enforcement on sales reads the true values.
/// Only non-None fields are updated (COALESCE keeps the existing value).
#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub fn update_customer_local(
    id: i64,
    name_ar: Option<String>,
    name_en: Option<String>,
    phone: Option<String>,
    vat_number: Option<String>,
    currency_code: Option<String>,
    credit_limit: Option<f64>,
    enforce_credit_limit: Option<bool>,
    payment_terms_days: Option<i64>,
    profile: Option<CustomerProfile>,
) -> Result<LocalCustomer, String> {
    update(id, name_ar, name_en, phone, vat_number, currency_code,
           credit_limit, enforce_credit_limit, payment_terms_days, profile.unwrap_or_default())
        .map_err(|e| e.to_string())
}

#[allow(clippy::too_many_arguments)]
fn update(
    id: i64,
    name_ar: Option<String>,
    name_en: Option<String>,
    phone: Option<String>,
    vat_number: Option<String>,
    currency_code: Option<String>,
    credit_limit: Option<f64>,
    enforce_credit_limit: Option<bool>,
    payment_terms_days: Option<i64>,
    profile: CustomerProfile,
) -> Result<LocalCustomer> {
    let conn = db::open()?;
    let enforce_int: Option<i64> = enforce_credit_limit.map(|b| b as i64);
    let include_int: Option<i64> = profile.include_in_statements.map(|b| b as i64);
    conn.execute(
        "UPDATE customers_local SET
           name_ar              = COALESCE(?2, name_ar),
           name_en              = COALESCE(?3, name_en),
           phone                = COALESCE(?4, phone),
           vat_number           = COALESCE(?5, vat_number),
           currency_code        = COALESCE(?6, currency_code),
           credit_limit         = COALESCE(?7, credit_limit),
           enforce_credit_limit = COALESCE(?8, enforce_credit_limit),
           payment_terms_days   = COALESCE(?9, payment_terms_days),
           cr_number              = COALESCE(?10, cr_number),
           email                  = COALESCE(?11, email),
           city                   = COALESCE(?12, city),
           district               = COALESCE(?13, district),
           street                 = COALESCE(?14, street),
           building_number        = COALESCE(?15, building_number),
           postal_code            = COALESCE(?16, postal_code),
           country                = COALESCE(?17, country),
           national_address_short = COALESCE(?18, national_address_short),
           location_lat           = COALESCE(?19, location_lat),
           location_lng           = COALESCE(?20, location_lng),
           location_link          = COALESCE(?21, location_link),
           include_in_statements  = COALESCE(?22, include_in_statements),
           branch_id              = CASE WHEN ?23 IS NULL THEN branch_id
                                         WHEN ?23 = 0 THEN NULL
                                         ELSE ?23 END,
           updated_at           = CURRENT_TIMESTAMP
         WHERE id = ?1",
        rusqlite::params![
            id, name_ar, name_en, phone, vat_number, currency_code,
            credit_limit, enforce_int, payment_terms_days,
            profile.cr_number, profile.email, profile.city, profile.district, profile.street,
            profile.building_number, profile.postal_code, profile.country, profile.national_address_short,
            profile.location_lat, profile.location_lng, profile.location_link,
            include_int, profile.branch_id,
        ],
    )?;
    let sql = format!("SELECT {SELECT_COLS} FROM customers_local WHERE id = ?1");
    conn.query_row(
        &sql,
        rusqlite::params![id],
        row_to_customer,
    ).map_err(Into::into)
}
